ipnsfs/system.go

```diff
@@ -0,0 +1,291 @@
+
+// KeyRoot represents the root of a filesystem tree pointed to by a given keypair
+type KeyRoot struct {
+	key ci.PrivKey
+
+	// node is the merkledag node pointed to by this keypair
+	node *dag.Node
+
+	// A pointer to the filesystem to access components
+	fs *Filesystem
+
+	// val represents the node pointed to by this key. It can either be a File or a Directory
+	val FSNode
+
+	repub *Republisher
+}
+
+func (fs *Filesystem) NewKeyRoot(parent context.Context, k ci.PrivKey) (*KeyRoot, error) {
+	hash, err := k.GetPublic().Hash()
+	if err != nil {
+		return nil, err
+	}
+
+	name := u.Key(hash).Pretty()
+
+	root := new(KeyRoot)
+	root.key = k
+	root.fs = fs
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	pointsTo, err := fs.nsys.Resolve(ctx, name)
+	if err != nil {
+		err = namesys.InitializeKeyspace(ctx, fs.dserv, fs.nsys, fs.pins, k)
+		if err != nil {
+			return nil, err
+		}
+
+		pointsTo, err = fs.nsys.Resolve(ctx, name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	mnode, err := fs.dserv.Get(pointsTo)
+	if err != nil {
+		return nil, err
+	}
+
+	root.node = mnode
+
+	root.repub = NewRepublisher(root, time.Millisecond*300, time.Second*3)
+	go root.repub.Run(parent)
+
+	pbn, err := ft.FromBytes(mnode.Data)
+	if err != nil {
+		log.Error("IPNS pointer was not unixfs node")
+		return nil, err
+	}
+
+	switch pbn.GetType() {
+	case ft.TDirectory:
+		root.val = NewDirectory(pointsTo.B58String(), mnode, root, fs)
+	case ft.TFile, ft.TMetadata, ft.TRaw:
+		fi, err := NewFile(pointsTo.B58String(), mnode, root, fs)
+		if err != nil {
+			return nil, err
+		}
+		root.val = fi
+	default:
+		panic("unrecognized! (NYI)")
+	}
+	return root, nil
+}
+
+func (kr *KeyRoot) GetValue() FSNode {
+	return kr.val
+}
+
+func (kr *KeyRoot) Open(tpath []string, mode int) (*File, error) {
+	if kr.val == nil {
+		// No entry here... what should we do?
+		panic("nyi")
+	}
+	if len(tpath) > 0 {
+		// Make sure our root is a directory
+		dir, ok := kr.val.(*Directory)
+		if !ok {
+			return nil, fmt.Errorf("no such file or directory: %s", tpath[0])
+		}
+
+		return dir.Open(tpath, mode)
+	}
+
+	switch t := kr.val.(type) {
+	case *Directory:
+		return nil, ErrIsDirectory
+	case *File:
+		return t, nil
+	default:
+		panic("unrecognized type, should not happen")
+	}
+}
+
+// closeChild implements the childCloser interface, and signals to the publisher that
+// there are changes ready to be published
+func (kr *KeyRoot) closeChild(name string, nd *dag.Node) error {
+	kr.repub.Touch()
+	return nil
+}
+
+// Publish publishes the ipns entry associated with this key
+func (kr *KeyRoot) Publish(ctx context.Context) error {
+	child, ok := kr.val.(FSNode)
+	if !ok {
+		return errors.New("child of key root not valid type")
+	}
+
+	nd, err := child.GetNode()
+	if err != nil {
+		return err
+	}
+
+	child.Lock()
+	k, err := kr.fs.dserv.Add(nd)
+	if err != nil {
+		child.Unlock()
+		return err
+	}
+	child.Unlock()
+	// Dont want to hold the lock while we publish
```

Comment:
because holding a lock while waiting on network IO will cause things to hang. Plus, it doesnt really matter if we are publishing multiple changes simultaneously. There are some funky conditions we can get into with one taking longer than the one following it, but that will be solved later on when we do some sort of versioning in ipns